File: electronbun/src/bun/flasher.ts
import { join } from "path";
import { tmpdir } from "os";
import { existsSync } from "fs";
import type { GithubRelease, FlashProgressPayload } from "../shared/types";

// ─── Constants ────────────────────────────────────────────────────────────────

const GITHUB_API_URL = "https://api.github.com/repos/LeoLTM/speedcamera/releases";
const GITHUB_USER_URL = "https://api.github.com/user";

// Only allow well-formed version tags from our own repo
const TAG_PATTERN = /^v\d+\.\d+\.\d+(-canary)?$/;

// ─── esptool binary path resolution ──────────────────────────────────────────
// Tries paths in order: packaged app, dev build layout, system PATH fallback

function resolveEsptoolPath(): string {
  const candidates = [
    join(import.meta.dir, "..", "resources", "esptool"),
    join(import.meta.dir, "..", "..", "resources", "esptool"),
  ];
  return candidates.find((p) => existsSync(p)) ?? "esptool";
}

const ESPTOOL_PATH = resolveEsptoolPath();

// ─── Helpers ──────────────────────────────────────────────────────────────────

function githubHeaders(token: string, accept = "application/vnd.github+json"): HeadersInit {
  const h: HeadersInit = { Accept: accept, "User-Agent": "speedcamera-app" };
  if (token) h["Authorization"] = `Bearer ${token}`;
  return h;
}

// ─── Push callback ────────────────────────────────────────────────────────────

type PushFn = (payload: FlashProgressPayload) => void;
let _push: PushFn | null = null;

/** Register the RPC push function (called once from index.ts after window creation). */
export function initFlasher(push: PushFn): void {
  _push = push;
}

function push(payload: FlashProgressPayload): void {
  _push?.(payload);
}

// ─── Active process handle ────────────────────────────────────────────────────

let _activeProc: ReturnType<typeof Bun.spawn> | null = null;

// ─── Public API ───────────────────────────────────────────────────────────────

/**
 * Verify that `token` has read access to the speedcamera releases.
 * Returns the GitHub login on success, or an error message on failure.
 */
export async function testGithubToken(
  token: string,
): Promise<{ valid: boolean; login?: string; error?: string }> {
  if (!token.trim()) {
    return { valid: false, error: "Token is empty" };
  }

  let resp: Response;
  try {
    resp = await fetch(`${GITHUB_API_URL}?per_page=1`, {
      headers: githubHeaders(token),
    });
  } catch (e) {
    return { valid: false, error: `Network error: ${e instanceof Error ? e.message : String(e)}` };
  }

  if (resp.status === 401) return { valid: false, error: "Invalid token — authentication failed" };
  if (resp.status === 403) return { valid: false, error: "Forbidden — check token scopes (needs repo read)" };
  if (resp.status === 404) return { valid: false, error: "Repository not found — check token permissions" };
  if (!resp.ok) return { valid: false, error: `GitHub API error: ${resp.status} ${resp.statusText}` };

  // Token works; also resolve the GitHub account name for confirmation
  let login: string | undefined;
  try {
    const userResp = await fetch(GITHUB_USER_URL, { headers: githubHeaders(token) });
    if (userResp.ok) {
      login = ((await userResp.json()) as { login: string }).login;
    }
  } catch {
    // Non-critical — token is still valid
  }

  return { valid: true, login };
}

/**
 * Fetch all releases for the speedcamera repo from the GitHub API.
 * Includes the API URL of the firmware.bin asset for each release so it can be
 * downloaded with auth (required for private repos).
 */
export async function listFirmwareReleases(token: string): Promise<GithubRelease[]> {
  const resp = await fetch(GITHUB_API_URL, { headers: githubHeaders(token) });
  if (resp.status === 401 || resp.status === 403) {
    throw new Error("GitHub authentication failed — check your token in the Firmware Update settings");
  }
  if (!resp.ok) {
    throw new Error(`GitHub API error: ${resp.status} ${resp.statusText}`);
  }

  const data = (await resp.json()) as Array<{
    tag_name: string;
    name: string;
    published_at: string;
    prerelease: boolean;
    assets: Array<{ name: string; url: string }>;
  }>;

  return data.map((r) => {
    const firmwareAsset = r.assets.find((a) => a.name === "firmware.bin");
    return {
      tag: r.tag_name,
      name: r.name,
      publishedAt: r.published_at,
      prerelease: r.prerelease,
      // API URL — download via this with Accept: application/octet-stream + auth header
      firmwareAssetApiUrl: firmwareAsset?.url ?? "",
    };
  });
}

/**
 * Download the firmware asset and flash it to the ESP8266 on the given port.
 *
 * Security:
 * - `releaseTag` is validated against TAG_PATTERN before use.
 * - `firmwareAssetApiUrl` must begin with the trusted GitHub API prefix.
 * - `port` comes from the serial port enumeration list (system-trusted).
 * - `Bun.spawn` receives an argument array — no shell interpretation.
 * - `ESPTOOL_PATH` is resolved at module load from a known bundle path.
 */
export async function flashFirmware(
  releaseTag: string,
  port: string,
  firmwareAssetApiUrl: string,
  token: string,
): Promise<void> {
  if (!TAG_PATTERN.test(releaseTag)) {
    throw new Error(`Invalid release tag: "${releaseTag}"`);
  }

  // Validate the asset URL is a known-safe GitHub API origin
  const ALLOWED_PREFIX = "https://api.github.com/repos/LeoLTM/speedcamera/releases/assets/";
  if (!firmwareAssetApiUrl.startsWith(ALLOWED_PREFIX)) {
    throw new Error("Invalid firmware asset URL");
  }

  const tmpPath = join(tmpdir(), `speedcamera-firmware-${releaseTag}.bin`);

  // ── Download via GitHub API asset endpoint (works for private repos) ───────
  push({ type: "downloading" });
  const dlResp = await fetch(firmwareAssetApiUrl, {
    headers: githubHeaders(token, "application/octet-stream"),
    redirect: "follow",
  });
  if (!dlResp.ok) {
    throw new Error(`Failed to download firmware: ${dlResp.status} ${dlResp.statusText}`);
  }
  await Bun.write(tmpPath, dlResp);

  // ── Spawn esptool ──────────────────────────────────────────────────────────
  // ESP8266 D1 Mini flash parameters:
  //   --flash_mode dout   required for D1 Mini (DOUT SPI mode)
  //   --flash_freq 40m    40 MHz clock
  //   --flash_size detect auto-detect flash size
  //   0x0                 write to start of flash
  const proc = Bun.spawn(
    [
      ESPTOOL_PATH,
      "--port", port,
      "--baud", "460800",
      "--chip", "esp8266",
      "write_flash",
      "--flash_mode", "dout",
      "--flash_freq", "40m",
      "--flash_size", "detect",
      "0x0",
      tmpPath,
    ],
    { stdout: "pipe", stderr: "pipe" },
  );
  _activeProc = proc;

  const decoder = new TextDecoder();
  let lineBuffer = "";

  function flushLines(chunk: string): void {
    lineBuffer += chunk;
    const parts = lineBuffer.split("\n");
    lineBuffer = parts.pop() ?? "";
    for (const line of parts) {
      const trimmed = line.replace(/\r$/, "").trim();
      if (trimmed) push({ type: "output", line: trimmed });
    }
  }

  async function drainStream(stream: ReadableStream<Uint8Array>): Promise<void> {
    const reader = stream.getReader();
    try {
      while (true) {
        const { done, value } = await reader.read();
        if (done) break;
        flushLines(decoder.decode(value, { stream: true }));
      }
    } finally {
      reader.releaseLock();
      flushLines(decoder.decode());
    }
  }

  await Promise.all([drainStream(proc.stdout), drainStream(proc.stderr)]);

  const exitCode = await proc.exited;
  _activeProc = null;

  if (exitCode !== 0) {
    throw new Error(`esptool exited with code ${exitCode}`);
  }

  push({ type: "done" });
}

/**
 * Kill the currently running esptool process (if any).
 */
export function cancelFlash(): void {
  if (_activeProc) {
    _activeProc.kill();
    _activeProc = null;
  }
}

File: headless-rust/src/integrations/flasher.rs
use crate::models::{FlashProgressPayload, GithubRelease};
use reqwest::header::{HeaderMap, HeaderValue, AUTHORIZATION, USER_AGENT};
use std::process::Stdio;
use tokio::io::{AsyncBufReadExt, BufReader};
use tokio::process::Command;
use tokio::sync::broadcast;

const GITHUB_API_URL: &str = "https://api.github.com/repos/LeoLTM/speedcamera/releases";
const GITHUB_USER_URL: &str = "https://api.github.com/user";

fn github_headers(token: &str) -> HeaderMap {
    let mut headers = HeaderMap::new();
    headers.insert(USER_AGENT, HeaderValue::from_static("speedcamera-rust"));
    if !token.is_empty() {
        if let Ok(val) = HeaderValue::from_str(&format!("Bearer {}", token)) {
            headers.insert(AUTHORIZATION, val);
        }
    }
    headers
}

pub async fn test_github_token(
    token: &str,
) -> Result<(bool, Option<String>), String> {
    if token.trim().is_empty() {
        return Ok((false, None));
    }

    let client = reqwest::Client::new();
    let res = client
        .get(GITHUB_USER_URL)
        .headers(github_headers(token))
        .send()
        .await
        .map_err(|e| e.to_string())?;

    if res.status().is_success() {
        let json: serde_json::Value = res.json().await.map_err(|e| e.to_string())?;
        let login = json["login"].as_str().map(|s| s.to_string());
        Ok((true, login))
    } else {
        Ok((false, None))
    }
}

pub async fn list_firmware_releases(token: &str) -> Result<Vec<GithubRelease>, String> {
    let client = reqwest::Client::new();
    let res = client
        .get(GITHUB_API_URL)
        .headers(github_headers(token))
        .send()
        .await
        .map_err(|e| e.to_string())?;

    if !res.status().is_success() {
        return Err(format!("GitHub API error: HTTP {}", res.status()));
    }

    let json: serde_json::Value = res.json().await.map_err(|e| e.to_string())?;
    let mut list = Vec::new();

    if let Some(releases) = json.as_array() {
        for r in releases {
            let tag = r["tag_name"].as_str().unwrap_or("").to_string();
            let name = r["name"].as_str().unwrap_or(&tag).to_string();
            let published_at = r["published_at"].as_str().unwrap_or("").to_string();
            let prerelease = r["prerelease"].as_bool().unwrap_or(false);

            let mut firmware_asset_api_url = String::new();
            if let Some(assets) = r["assets"].as_array() {
                for a in assets {
                    let asset_name = a["name"].as_str().unwrap_or("");
                    if asset_name.ends_with(".bin") || asset_name.contains("firmware") {
                        firmware_asset_api_url = a["url"].as_str().unwrap_or("").to_string();
                        break;
                    }
                }
            }

            list.push(GithubRelease {
                tag,
                name,
                published_at,
                prerelease,
                firmware_asset_api_url,
            });
        }
    }

    Ok(list)
}

pub async fn flash_firmware(
    _release_tag: &str,
    port: &str,
    firmware_asset_api_url: &str,
    token: &str,
    tx: broadcast::Sender<FlashProgressPayload>,
) {
    let _ = tx.send(FlashProgressPayload::Downloading);

    let client = reqwest::Client::new();
    let mut req = client.get(firmware_asset_api_url).headers(github_headers(token));
    req = req.header("Accept", "application/octet-stream");

    let bytes = match req.send().await {
        Ok(res) if res.status().is_success() => match res.bytes().await {
            Ok(b) => b,
            Err(e) => {
                let _ = tx.send(FlashProgressPayload::Error {
                    message: format!("Failed to read binary: {}", e),
                });
                return;
            }
        },
        Ok(res) => {
            let _ = tx.send(FlashProgressPayload::Error {
                message: format!("Failed to download firmware asset: HTTP {}", res.status()),
            });
            return;
        }
        Err(e) => {
            let _ = tx.send(FlashProgressPayload::Error {
                message: format!("Download network error: {}", e),
            });
            return;
        }
    };

    let temp_file = std::env::temp_dir().join("speedcamera_firmware.bin");
    if let Err(e) = std::fs::write(&temp_file, &bytes) {
        let _ = tx.send(FlashProgressPayload::Error {
            message: format!("Failed to write temp firmware file: {}", e),
        });
        return;
    }

    let esptool_cmd = "esptool.py";
    let mut child = match Command::new(esptool_cmd)
        .arg("--chip")
        .arg("esp8266")
        .arg("--port")
        .arg(port)
        .arg("--baud")
        .arg("460800")
        .arg("write_flash")
        .arg("0x0")
        .arg(temp_file.to_str().unwrap_or(""))
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .spawn()
    {
        Ok(c) => c,
        Err(_) => {
            // Try standard esptool
            match Command::new("esptool")
                .arg("--chip")
                .arg("esp8266")
                .arg("--port")
                .arg(port)
                .arg("--baud")
                .arg("460800")
                .arg("write_flash")
                .arg("0x0")
                .arg(temp_file.to_str().unwrap_or(""))
                .stdout(Stdio::piped())
                .stderr(Stdio::piped())
                .spawn()
            {
                Ok(c) => c,
                Err(e) => {
                    let _ = tx.send(FlashProgressPayload::Error {
                        message: format!("esptool not found or failed to spawn: {}", e),
                    });
                    return;
                }
            }
        }
    };

    if let Some(stdout) = child.stdout.take() {
        let mut reader = BufReader::new(stdout).lines();
        let tx_clone = tx.clone();
        tokio::spawn(async move {
            while let Ok(Some(line)) = reader.next_line().await {
                let _ = tx_clone.send(FlashProgressPayload::Output { line });
            }
        });
    }

    match child.wait().await {
        Ok(status) if status.success() => {
            let _ = tx.send(FlashProgressPayload::Done);
        }
        Ok(status) => {
            let _ = tx.send(FlashProgressPayload::Error {
                message: format!("esptool exited with status {}", status),
            });
        }
        Err(e) => {
            let _ = tx.send(FlashProgressPayload::Error {
                message: format!("esptool execution error: {}", e),
            });
        }
    }
}
